### PR TITLE
1000 bulk delete

### PR DESCRIPTION
**Pull request by garethbowen**
Bit of a big one I'm afraid. I'm planning to review the icons used for the buttons, so don't worry about those. There is also a bit of a performance issue with Select All which might have to fetch 1000s of docs, but I've split this off into #2302. Everything else should be good to go. Issue: #1000

### CODE REVIEW

File: static/js/controllers/confirm-modal.js [hunk 1]
```diff
@@ -81,21 +90,13 @@
...
             return;
           }
 
           // Simple true/false result.
-          if (result) {
-            $scope.ok();
-          } else {
-            _setErrorMode();
-          }
+          $scope.ok();
```
Comment (estellecomment) [inline static/js/controllers/confirm-modal.js]: That removes some logic : if processing function is synchronous, it returns a boolean, that indicates whether processing worked (saving, typically)
Did you remove it on purpose?

File: static/js/controllers/inbox.js [hunk 2]
```diff
@@ -660,6 +661,10 @@ var feedback = require('../modules/feedback'),
           });
       };
 
+      $scope.setSelectMode = function(value) {
```
Comment (estellecomment) [inline static/js/controllers/inbox.js]: comment explaining what selectMode corresponds to in the UI would be nice.

File: static/js/controllers/confirm-modal.js (hunk 1 above)
Comment (garethbowen) [inline static/js/controllers/confirm-modal.js]: Right, but in line 85 (of the changed version) we've already dealt with the processing function being synchronous and returning false. I can't see any code that would change the value of result between that line and here. Maybe I'm missing something.

File: static/js/controllers/reports.js [hunk 3]
```diff
@@ -106,23 +126,38 @@ var _ = require('underscore'),
         $scope.displayFields = fields;
       };
 
+      var setActionBar = function() {
+        var model = {
+          doc: _.pluck($scope.selected, 'report')
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: this is a list, should be called `docs`

File: static/js/controllers/reports.js [hunk 4]
```diff
@@ -11,11 +11,31 @@ var _ = require('underscore'),
...
+      SearchFilters,
+      Settings,
+      TranslateFrom,
+      Verified
+    ) {
+      'ngInject';
 
       $scope.selectedGroup = null;
-      $scope.selected = null;
+      $scope.selected = [];
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: can you document here the expected format for `$scope.selected`?

File: static/js/controllers/reports.js [hunk 5]
```diff
@@ -106,23 +126,38 @@ var _ = require('underscore'),
...
+          setActionBar();
+        } else {
+          liveList.setSelected(doc._id);
+
+          var refreshing = doc &&
+                           $scope.selected.length &&
+                           $scope.selected[0].report._id === doc._id;
+          $scope.selected = [ { report: doc, expanded: true } ];
+          setTitle(doc);
+          setActionBar();
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: why is there no args any more?

File: static/js/controllers/reports.js [hunk 6]
```diff
@@ -151,8 +186,9 @@ var _ = require('underscore'),
           $scope.clearSelected();
           return;
         }
-
-        $scope.clearSelected();
+        if (!$scope.selectMode) {
+          $scope.clearSelected();
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: can you document why clearSelected isn't used in selectMode?

File: static/js/controllers/reports-content.js [hunk 7]
```diff
@@ -45,10 +47,34 @@
...
+            }
+          }
+        }
+      };
+
       Changes({
         key: 'reports-content',
         filter: function(change) {
-          return $scope.selected && $scope.selected._id === change.id;
+          return $scope.selected &&
```
Comment (estellecomment) [inline static/js/controllers/reports-content.js]: General thought : is there any time where `$scope.selected` should be null? Now that it's a list, it should be empty but not null, right? So I don't think we want that case to fail silently... Here and kindof everywhere. 
It could be best to test for `$scope.selected.length` only, and crash if there's no `selected`, and we'll find bugs that way.

Comment (garethbowen) [inline static/js/controllers/reports-content.js]: It probably will never be null, but there _might_ be a race condition where the Changes event fires as the controller is destroying itself. Because the destroy handling code and this code are both event based I don't fully trust that the controller will clean up in a deterministic fashion.

File: static/js/controllers/reports.js (hunk 3 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: Good call. I've renamed it to "selected" as this is more descriptive than "doc" or "docs".

File: static/js/controllers/reports.js [hunk 8]
```diff
@@ -11,11 +11,35 @@ var _ = require('underscore'),
...
+      Settings,
+      TranslateFrom,
+      Verified
+    ) {
+      'ngInject';
+
+      // selected objects have the form
+      //    { report: { ... }, expanded: false }
+      // where the report is the db doc and expanded is whether to
+      // show the details or just the summary in the content pane.
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: ❤️

File: static/js/controllers/reports.js [hunk 9]
```diff
@@ -92,36 +116,68 @@ var _ = require('underscore'),
...
-          _id: doc._id,
-          verified: doc.verified,
-          type: doc.content_type,
-          sendTo: doc
-        });
+      $scope.setSelected = function(doc) {
+        var refreshing = true;
+        var displayFields = getDisplayFields(doc);
+        if ($scope.selectMode) {
+          var existing = _.findWhere($scope.selected, { _id: doc._id });
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: `selected` is like `[{ report: { _id: 'aaa' }, expanded: false }, { report: { _id: 'bbb' }, expanded: false }]`
I don't think `_.findWhere` looks deeper than first level of attributes.
So I don't think `existing` is ever non-null!

File: static/js/controllers/reports.js [hunk 10]
```diff
@@ -131,34 +187,60 @@ var _ = require('underscore'),
...
-            })
+            _setSelected(doc[0]);
+          })
           .catch(function(err) {
             $log.error('Error fetching formatted report', err);
           });
       };
 
+      $scope.deselectReport = function(report) {
+        for (var i = 0; i < $scope.selected.length; i++) {
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: whoa iterating instead of underscore, don't remember what it is any more, hurts brain, can't read

File: static/js/controllers/reports.js [hunk 11]
```diff
@@ -131,34 +187,60 @@ var _ = require('underscore'),
...
+          if ($scope.selected[i]._id === report._id) {
+            $scope.selected.splice(i, 1);
+          }
+        }
+        $('#reports-list li[data-record-id="' + report._id + '"] input[type="checkbox"]')
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: could you extract into a function `makeChecked(id, bool)`? Isolating the jquerying into fewer places would be nicer.

File: static/js/controllers/reports.js [hunk 12]
```diff
@@ -131,34 +187,60 @@ var _ = require('underscore'),
...
-              _setSelected(doc[0]);
-            })
+            _setSelected(doc[0]);
+          })
           .catch(function(err) {
             $log.error('Error fetching formatted report', err);
           });
       };
 
+      $scope.deselectReport = function(report) {
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: random thought, does this behave nicely if you call it in non-select mode?

File: static/js/controllers/reports.js (hunk 9 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: In the latest version an entry in selected is actually more like: `{ _id: 'aaa', report: {}, summary: {}, expanded: false, loading: false}`. I'll update the documentation.

File: static/js/controllers/reports.js (hunk 10 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: I wish underscore returned the index of the found item...

File: static/js/controllers/reports.js [hunk 13]
```diff
@@ -131,34 +187,60 @@ var _ = require('underscore'),
...
-        $scope.clearSelected();
         $scope.setLoadingContent(report);
-
+        if (!$scope.selectMode) {
+          // in selected mode we append to the list so don't clear it
+          $scope.clearSelected();
+        }
         _fetchFormattedReport(report)
+          .then(function(formatted) {
+            return formatted && formatted.length && formatted[0];
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: just `if(long condition)` in the `then` block below would be more concise (but this works too)

File: static/js/controllers/reports.js (hunk 12 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: I guess it would... but that shouldn't be possible in the UI.

File: static/js/controllers/reports.js [hunk 14]
```diff
@@ -250,17 +335,18 @@ var _ = require('underscore'),
...
           }
         }
-
       };
 
       $scope.$on('ClearSelected', function() {
-        $scope.selected = null;
+        $scope.selected = [];
+        $('#reports-list input[type="checkbox"]')
+          .prop('checked', false);
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: same here, `makeAllChecked(bool)` would look nice

File: static/js/controllers/reports.js (hunk 11 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: This is the only place I can see that checks by ID so I don't think this would gain anything. I'd love to use angular for everything but anything in the LHS has to be manually managed...

File: static/js/controllers/reports.js [hunk 15]
```diff
@@ -250,17 +335,18 @@ var _ = require('underscore'),
...
         liveList.clearSelected();
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: Hey, should the livelist selection always match the checkbox selection? If yes, how could we wrap them together so that they always happen together?

Comment (garethbowen) [inline static/js/controllers/reports.js]: The livelist clearSelected method is type agnostic so I didn't want to put report specific code in there. When other tabs have a bulk delete/edit option this would make sense.

File: static/js/controllers/reports.js (hunk 10 above)
Comment (estellecomment) [inline static/js/controllers/reports.js]: `_.findIndex` ?

Comment (garethbowen) [inline static/js/controllers/reports.js]: Whoa - that's new to me! Nice.

File: static/js/controllers/reports.js [hunk 16]
```diff
@@ -250,17 +336,18 @@ var _ = require('underscore'),
             _query();
           }
         }
-
       };
 
       $scope.$on('ClearSelected', function() {
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: could this just call `$scope.clearSelected()`?

File: static/js/controllers/reports.js [hunk 17]
```diff
@@ -369,13 +456,78 @@ var _ = require('underscore'),
         });
       };
       $scope.resetFilterModel = function() {
+        if ($scope.selectMode && $scope.selected && $scope.selected.length) {
+          // can't filter when in select mode
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: can we log this as an error, since it shouldn't be happening?

File: static/js/controllers/reports.js (hunk 16 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: No. $scope.clearSelected emits this event so child controllers can clean up their context, so doing that would cause an infinite loop...

File: static/js/controllers/reports.js [hunk 18]
```diff
@@ -369,13 +456,78 @@ var _ = require('underscore'),
...
+          var reportId = target.attr('data-record-id');
+          var checkbox = target.find('input[type="checkbox"]');
+          var alreadySelected = _.findWhere($scope.selected, { _id: reportId });
+          $timeout(function() {
+            checkbox.prop('checked', !alreadySelected);
+            if (!alreadySelected) {
+              $scope.selectReport(reportId);
+            } else {
+              for (var i = 0; i < $scope.selected.length; i++) {
+                if ($scope.selected[i]._id === reportId) {
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: reuse `deselectReport` here

File: static/js/controllers/reports.js [hunk 19]
```diff
@@ -369,13 +456,78 @@ var _ = require('underscore'),
...
 
+      $('.inbox').on('click', '#reports-list .message-wrapper', function(e) {
+        if ($scope.selectMode) {
+          e.preventDefault();
+          e.stopPropagation();
+          var target = $(e.target).closest('li');
+          var reportId = target.attr('data-record-id');
+          var checkbox = target.find('input[type="checkbox"]');
+          var alreadySelected = _.findWhere($scope.selected, { _id: reportId });
+          $timeout(function() {
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: for my information, why the timeout here?

Comment (garethbowen) [inline static/js/controllers/reports.js]: Good question. I don't remember. I'll try taking it out. It may have been for debugging purposes.

File: static/js/controllers/reports.js (hunk 16 above)
Comment (estellecomment) [inline static/js/controllers/reports.js]: oh, I see, `$scope.clearSelected` belongs to inbox. Got it.
I should get around to https://github.com/medic/medic-webapp/issues/2133 so that I stop doing these thought loops over and over

File: static/js/controllers/reports-content.js (hunk 7 above)
Comment (estellecomment) [inline static/js/controllers/reports-content.js]: Interesting. Ok.
But everywhere else in the file though?

File: static/js/controllers/reports.js (hunk 15 above)
Comment (estellecomment) [inline static/js/controllers/reports.js]: You could do a wrapper, in this class. The wrapper would forward calls to underlying livelist, and do the checkboxing.
Also, it's not only for `clearSelected`, it's for all operations.

Comment (garethbowen) [inline static/js/controllers/reports.js]: Because the livelist functions called and the selectors used to check the checkboxes are almost always different for the different invocations you'd end up with something like...
```
function modifySelection(liveListFunction, checkboxSelector, checked) {
  liveListFunction();
  $(checkboxSelector).prop('checked', checked);
}
```

This doesn't reduce complexity at all because all the interesting code is still in the calling function, and it just adds yet another layer of abstraction. The one benefit is it'll remind people to do both.

File: static/js/controllers/reports.js (hunk 18 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: Done.

File: static/js/controllers/reports.js (hunk 19 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: I remember now. If you click the checkbox then this code fires, we set the checkbox to checked, then the user action fires and unchecks it. By delaying we are guaranteed to set the checkbox to the correct value no matter what the user did. I've added a comment to make this clearer.

File: static/js/controllers/reports.js (hunk 15 above)
Comment (estellecomment) [inline static/js/controllers/reports.js]: Yeah, there are a bunch of functions, it does add a bunch of code, it doesn't reduce it. But it's cleaner, better organized code, easier to understand and modify. It does add a layer of abstraction, and the point of abstraction is to do exactly that!

There's basically select/unselect by id, select/unselect all, that's 3 functions, it's not long to write out fully. (am I being optimistic? :D)

While you're at it you could also package in the modification of `$scope.selected` (unless there's a reason it shouldn't match the livelist selection and the checkboxes).

File: static/js/controllers/reports.js [hunk 20]
```diff
@@ -480,6 +480,8 @@ var _ = require('underscore'),
           var reportId = target.attr('data-record-id');
           var checkbox = target.find('input[type="checkbox"]');
           var alreadySelected = _.findWhere($scope.selected, { _id: reportId });
+          // timeout so if the user clicked the checkbox it has time to
+          // register before we set it to the correct value.
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: interesting, what needs time to register? curious

File: static/js/services/delete-doc.js [hunk 21]
```diff
@@ -1,66 +1,65 @@
...
+
 (function () {
 
   'use strict';
 
   var inboxServices = angular.module('inboxServices');
 
-  inboxServices.factory('DeleteDoc', ['DB',
-    function(DB) {
+  inboxServices.factory('DeleteDoc',
```
Comment (estellecomment) [inline static/js/services/delete-doc.js]: can you rename to `DeleteDocs`, plural?

File: static/js/services/delete-doc.js [hunk 22]
```diff
@@ -1,66 +1,65 @@
...
-          .get(docId)
-          .then(function(doc) {
-            updateParent(doc, function(err) {
-              if (err) {
-                return callback(err);
+      /**
+       * Delete the given docs. If 'person' type then also fix the
+       * contact hierarchy.
+       *
+       * @param docs {Object|Array} Document or array of documents to delete.
```
Comment (estellecomment) [inline static/js/services/delete-doc.js]: 😍  @param annotation

File: static/js/services/delete-doc.js [hunk 23]
```diff
@@ -1,66 +1,65 @@
...
+          docs = [ docs ];
+        }
+        var toUpdate = docs.map(function(doc) {
+          return {
+            _id: doc._id,
+            _rev: doc._rev,
+            _deleted: true
+          };
+        });
+        return $q.all(docs.map(function(doc) {
```
Comment (estellecomment) [inline static/js/services/delete-doc.js]: Hmm what if you are running `DeleteDocs([doc, doc.parent])`? 
Then `doc.parent` is being edited twice in parallel, right? Is it going to be consistent, been the get and the put? Are we saved by the single-threadedness?

File: static/js/controllers/inbox.js [hunk 24]
```diff
@@ -607,37 +615,49 @@ var feedback = require('../modules/feedback'),
...
-      // TODO promisify DeleteDoc
-      var _deleteDoc = function(id) {
-        var deferred = $q.defer();
-        if (!id) {
-          return deferred.reject('Error deleting document : no docToDeleteId set');
+      $scope.deleteDoc = function(docs) {
+        if (!docs) {
+          return;
+        }
+        if (!_.isArray(docs)) {
```
Comment (estellecomment) [inline static/js/controllers/inbox.js]: that's dealt with in `DeleteDoc` already

File: static/js/controllers/inbox.js [hunk 25]
```diff
@@ -607,37 +615,49 @@ var feedback = require('../modules/feedback'),
...
+          docs = [ docs ];
+        }
+        if (!docs.length) {
```
Comment (estellecomment) [inline static/js/controllers/inbox.js]: that would be taken care of by the `if (!docs)` above, no?

File: static/js/controllers/messages-content.js [hunk 26]
```diff
@@ -180,7 +180,7 @@ var _ = require('underscore'),
         filter: function(change) {
           return $scope.currentTab === 'messages' &&
             $scope.selected &&
-            $scope.selected.id === change.id;
+            _.findWhere($scope.selected.messages, { id: change.id });
```
Comment (estellecomment) [inline static/js/controllers/messages-content.js]: is this related to this issue or did it sneak in?

File: static/js/controllers/reports-content.js [hunk 27]
```diff
@@ -45,15 +47,45 @@
         setMessageState(group, 'muted', 'scheduled');
       };
 
+      $scope.toggleExpand = function(selection) {
+        if ($scope.selectMode) {
```
Comment (estellecomment) [inline static/js/controllers/reports-content.js]: better: `if (!selectMode) {return}`

File: static/js/controllers/reports-content.js [hunk 28]
```diff
@@ -45,15 +47,45 @@
         setMessageState(group, 'muted', 'scheduled');
       };
 
+      $scope.toggleExpand = function(selection) {
```
Comment (estellecomment) [inline static/js/controllers/reports-content.js]: selection is a single report, right? Call it `report` so it's clearer

File: static/js/controllers/reports-content.js [hunk 29]
```diff
@@ -45,15 +47,45 @@
...
+        if ($scope.selectMode) {
+          if (selection.report || selection.expanded) {
+            selection.expanded = !selection.expanded;
+          } else {
```
Comment (estellecomment) [inline static/js/controllers/reports-content.js]: when does this happen? Does the selection object not always have `report` and `expanded` fields?

File: static/js/controllers/reports.js (hunk 15 above)
Comment (estellecomment) [inline static/js/controllers/reports.js]: Additional point : that piece of checkboxing is alone in livelist: 
https://github.com/medic/medic-webapp/blob/cfae7ff81efb4114a1ac222a667080d83e88f158/static/js/services/live-list.js#L96

You could pull it out of livelist and into the wrapper, and thus do all the checkboxing in the same place.

File: static/js/services/live-list.js [hunk 30]
```diff
@@ -330,8 +337,8 @@ angular.module('inboxServices').factory('LiveList',
     }
 
     function _update(listName, updatedItem) {
-      _remove(listName, updatedItem);
-      _insert(listName, updatedItem);
+      var removed = _remove(listName, updatedItem);
```
Comment (estellecomment) [inline static/js/services/live-list.js]: interesting. Why is this new, when does this happen?

File: static/js/services/update-user.js [hunk 31]
```diff
@@ -127,8 +127,8 @@ var _ = require('underscore');
     }
   ]);
 
-  inboxServices.factory('DeleteUser', ['$cacheFactory', '$http', 'DeleteDoc',
-    function($cacheFactory, $http, DeleteDoc) {
+  inboxServices.factory('DeleteUser', ['$cacheFactory', '$http', 'DB', 'DeleteDoc',
+    function($cacheFactory, $http, DB, DeleteDoc) {
```
Comment (estellecomment) [inline static/js/services/update-user.js]: optional : switch to `nginject` dependencies

File: templates/partials/filters/date.html [hunk 32]
```diff
@@ -1,5 +1,5 @@
-<span id="dateRangeDropdown" class="filter dropdown mm-dropdown">
-  <span id="date-filter" class="mm-button">
+<span id="dateRangeDropdown" class="filter dropdown mm-dropdown" ng-class="{ 'disabled': selectMode && selected.length }">
```
Comment (estellecomment) [inline templates/partials/filters/date.html]: I think the disabling would make more sense in `reports_fitlers.html`.
And you could do them all in the same file, which would be nice.

File: templates/partials/sender.html [hunk 33]
```diff
@@ -4,10 +4,10 @@
     ng-bind="message.contact.name || (!message.form && message.name) || message.from || message.sent_by"></span>
     <span class="phone"
     ng-hide="message.name && message.name === message.facility.contact.phone"
-    ng-bind="message.contact.phone"></span>
+    ng-bind="message.contact.phone || message.phone"></span>
```
Comment (estellecomment) [inline templates/partials/sender.html]: same here, did that sneak in from another issue?

File: tests/karma/unit/controllers/inbox.js [hunk 34]
```diff
@@ -160,7 +157,6 @@ describe('InboxCtrl controller', function() {
       chai.assert(spyState.go.called, 'Should change state');
       chai.expect(spyState.go.args[0][0]).to.equal(spyState.current.name);
       chai.expect(spyState.go.args[0][1]).to.deep.equal({ id: null });
-      chai.assert(snackbar.called, 'Should display toast');
```
Comment (estellecomment) [inline tests/karma/unit/controllers/inbox.js]: snackbar gone?

File: tests/karma/unit/services/delete-doc.js [hunk 35]
```diff
@@ -76,22 +104,21 @@ describe('DeleteDoc service', function() {
...
+      chai.expect(bulkDocs.args[0][0].length).to.equal(2);
+      chai.expect(bulkDocs.args[0][0][0]._id).to.equal(person._id);
+      chai.expect(bulkDocs.args[0][0][0]._deleted).to.equal(true);
+      chai.expect(bulkDocs.args[0][0][1]._id).to.equal(clinic._id);
+      chai.expect(bulkDocs.args[0][0][1].contact).to.equal(null);
     });
   });
 
-  it('done update clinic when phone does not match', function(done) {
+  it('does not update clinic when phone does not match', function() {
```
Comment (estellecomment) [inline tests/karma/unit/services/delete-doc.js]: interesting. Why do we do that?

File: static/js/controllers/reports.js (hunk 20 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: When a user clicks on a checkbox the browser toggles the checkbox state. If I go ahead and set checked to true then the browser will toggle it immediately afterwards. Instead I wait for the browser to do its thing then set it to what I want. The alternative is to register a second event listener on the checkbox itself which also does this code but also prevents default. I opted for the option which means fewer event listeners.

File: static/js/services/delete-doc.js (hunk 23 above)
Comment (garethbowen) [inline static/js/services/delete-doc.js]: There isn't a `get` step any more so that would give a document update conflict. This isn't possible with the current design as you can't bulk delete contacts. We should revisit it later.

Contact deletion in general needs to be well thought through. For example if you're deleting parent should that automatically delete the children or leave them as orphans? Too hard for now...

File: static/js/controllers/inbox.js (hunk 24 above)
Comment (garethbowen) [inline static/js/controllers/inbox.js]: Good point. Fixing.

Comment (garethbowen) [inline static/js/controllers/inbox.js]: Actually looking at the code we use `docs` outside of DeleteDoc as well for working out what to show in the modal, eg: "delete 2 docs?". Without the isArray check here we'd have to jump through hoops in the other two places.

File: static/js/controllers/inbox.js (hunk 25 above)
Comment (garethbowen) [inline static/js/controllers/inbox.js]: No. If you pass in `[]` then this condition will be true.

File: static/js/controllers/messages-content.js (hunk 26 above)
Comment (garethbowen) [inline static/js/controllers/messages-content.js]: I think it's a separate issue. Sorry about that.

File: static/js/controllers/reports-content.js (hunk 27 above)
Comment (garethbowen) [inline static/js/controllers/reports-content.js]: Fair enough. Fixed.

File: static/js/controllers/reports-content.js (hunk 29 above)
Comment (garethbowen) [inline static/js/controllers/reports-content.js]: The search performance changes introduced a `summary` property which has all the data needed for the collapsed version. The `report` is only populated on first expand but then it's cached for future expansions.

The `else` is all about loading the `report` property, but if the `report` property is already populated or we're collapsing the row (eg: `selection.expanded === true`) then don't worry about loading the report.

File: static/js/services/live-list.js (hunk 30 above)
Comment (garethbowen) [inline static/js/services/live-list.js]: It's not exactly new. I've just changed the existing remove function to return the element that was removed, and pass it in to the insert function.

In the case of reports this means I can write code which sets the state of the checkbox of the new `li` element to the same as the old checkbox. This is fired for example when the changes feed notices an element in the list has changed. In this case we want the selection to remain the same but the doc content to be updated.

This is what fixed the bug you noticed where selections could get out of sync. In your case I suspect selecting the item marked it as read which fired the change listener which updated the $scope.selected just fine, but the checkbox was cleared when the livelist updated.

File: static/js/services/update-user.js (hunk 31 above)
Comment (garethbowen) [inline static/js/services/update-user.js]: Next time...

File: templates/partials/filters/date.html (hunk 32 above)
Comment (garethbowen) [inline templates/partials/filters/date.html]: That was my initial plan too, but adding the class to that specific element means we get some of the disabling for free from bootstrap. If we put it on the parent element we'd have to write more css to implement our own disabling.

File: templates/partials/sender.html (hunk 33 above)
Comment (garethbowen) [inline templates/partials/sender.html]: No this is relevant to this one. Specifically it's so we can render the report summary or full details for performance reasons. See #2302

File: tests/karma/unit/services/delete-doc.js (hunk 35 above)
Comment (garethbowen) [inline tests/karma/unit/services/delete-doc.js]: The test rename? I assumed the old one was a typo as it didn't make sense to me.

Removing the `done`? When testing a promise based service if you return the promise then karma uses that for done so you don't have to check yourself. It also fails the test on `catch`. So it saves us a bit of boilerplate in the tests.

File: tests/karma/unit/controllers/inbox.js (hunk 34 above)
Comment (garethbowen) [inline tests/karma/unit/controllers/inbox.js]: Ugh. Snackbar is still there and works perfectly. But because we can't mock $translate (don't get me started) the call never returns in unit testing.

I'm seriously wondering if we need to wrap `$translate` in our own translate service, just so we can mock it.

File: static/js/services/live-list.js (hunk 30 above)
Comment (estellecomment) [inline static/js/services/live-list.js]: Cool.
Could you just pass a `bool` `isChecked`, instead of passing the whole removedElement? (unless I'm missing something)

File: static/js/controllers/reports.js [hunk 36]
```diff
@@ -369,13 +461,74 @@ var _ = require('underscore'),
...
+      $('.inbox').on('click', '#reports-list .message-wrapper', function(e) {
+        if ($scope.selectMode) {
+          e.preventDefault();
+          e.stopPropagation();
+          var target = $(e.target).closest('li');
+          var reportId = target.attr('data-record-id');
+          var checkbox = target.find('input[type="checkbox"]');
+          var alreadySelected = _.findWhere($scope.selected, { _id: reportId });
+          // timeout so if the user clicked the checkbox it has time to
+          // register before we set it to the correct value.
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: Sorry still don't get this. What needs time to do what?

File: templates/modals/delete_doc_confirm.html [hunk 37]
```diff
@@ -1,17 +1,18 @@
 <div role="document" id="delete-confirm">
   <div class="modal-header">
     <button type="button" class="cancel close" ng-click="cancel()">&times;</button>
-    <h3>{{'Confirm' | translate}}</h3>
+    <h3 translate>Confirm</h3>
   </div>
   <div class="modal-body">
-    <p>{{'confirm.delete' | translate}}</p>
+    <p ng-show="model.docs.length === 1" translate>confirm.delete</p>
+    <p ng-show="model.docs.length !== 1" translate translate-values="{ number: {{model.docs.length}} }">confirm.delete.plural</p>
```
Comment (estellecomment) [inline templates/modals/delete_doc_confirm.html]: you actually only need `docs.length` in the modal. Can you pass `model.numDocs` instead of `model.docs`?

File: static/js/controllers/reports.js [hunk 38]
```diff
@@ -194,14 +282,17 @@ var _ = require('underscore'),
             $scope.error = false;
             $scope.errorSyntax = false;
             _updateLiveList(data);
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: instead of using `syncCheckboxes` below, I'd add the true/false value in the `data` here, and pass it to `_updateLiveList`, which would set the right values directly.

Comment (estellecomment) [inline static/js/controllers/reports.js]: also, it's not from this PR changes, but l.69 (I can't add comment there because it's not edited :( ), there's `liveList.update(report, false)`, and I think the second arg is ignored. I don't know what it's meant to be.

But to pass the checked status, you could do `liveList.update(report, isChecked)`.

File: static/js/controllers/reports.js [hunk 39]
```diff
@@ -131,34 +188,65 @@ var _ = require('underscore'),
...
-              _setSelected(doc[0]);
-            })
+            _setSelected(doc[0]);
+          })
           .catch(function(err) {
             $log.error('Error fetching formatted report', err);
           });
       };
 
+      var spliceSelected = function(id) {
```
Comment (estellecomment) [inline static/js/controllers/reports.js]: rename to `_spliceSelected`

Comment (estellecomment) [inline static/js/controllers/reports.js]: also `getNextHalfHour` -> `_getNextHalfHour`, `initEditMessageModal`, `deselectAll`, `setTitle`, `getFields`... 
Unless I'm not getting the notation.

File: tests/karma/unit/services/delete-doc.js (hunk 35 above)
Comment (estellecomment) [inline tests/karma/unit/services/delete-doc.js]: Sorry, I wasn't clear. 
I mean, why do we not update the clinic when the phone doesn't match? (I could maybe leave such questions out of the code review, which has enough comments as it is :D )

Comment (garethbowen) [inline tests/karma/unit/services/delete-doc.js]: Not sure. Maybe it's some sort of cheap optimistic locking.

File: static/js/controllers/reports.js (hunk 36 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: When a user clicks on a checkbox the browser toggles it between checked and unchecked. This is bad in this case because we want to control the checking and unchecking. One solution is to let the browser do its thing and then immediately set it the way we want it set. This is achieved by letting the event complete and then running our checked state setting code. The "timeout" is putting our code at the end of the execution queue so it'll run immediately after the browser has finished doing the wrong thing.

File: templates/modals/delete_doc_confirm.html (hunk 37 above)
Comment (garethbowen) [inline templates/modals/delete_doc_confirm.html]: Yeah we could just pass the length rather than the array. I don't think it'll make any difference to the JS engine. Stylistically I'm happy with either.

File: static/js/controllers/reports.js (hunk 38 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: I really don't want LiveList to have to know too much about how Reports works because that's scope creep IMO. Ideally LiveList would deal with rendering the items but not setting state (checking checkboxes, etc). I've broken this rule a little where necessary but I really want to keep this to a minimum. Otherwise all logic/state/etc will end up in ListList and it'll be a great mess.

I've fixed the ignored param.

File: static/js/controllers/reports.js (hunk 39 above)
Comment (garethbowen) [inline static/js/controllers/reports.js]: Heh. Yeah. There's inconsistency here. Initially I was using `_` to indicate private or internal functions, but then I realised that anything that's not defined on `$scope` is private or internal so it was unnecessary and confusing. I'll remove `_` from all functions in this controller.
